#27015 gemma3-27b bringup on T3K

handrewsTT opened the pull request:
### Ticket
#27015

### Problem description
Model bringup for gemma3 27b on T3K

### What's changed
Some changes to TTT modules to support gemma 3 model including:
- fix bug in attention.py where wrong key name was used
- add rope matrix functions to common.py for tests 
- add multimodal prompt encoder to common.py 
- add pre and post MLP layer norm if defined in model config to decoder.py
- add non-llama specific generator forward funcs for multimodal prefill and decode in generator.py
- update vllm integration in generator_vllm.py
- add configurable lm head dtypes and memory configs in lm_head.py
- add MLP activation setting in mlp.py


Tests for gemma 3 have been added to the following workflows:
- .github/workflows/fast-dispatch-full-regressions-and-models-impl.yaml (gemma-3-4b-it N150/N300 perf)
- .github/workflows/single-card-demo-tests-impl.yaml (gemma-3-4b-it N150/N300 functional)
- .github/workflows/t3000-demo-tests-impl.yaml (gemma-3-27b-it T3K perf)
- .github/workflows/t3000-perplexity-tests-impl.yaml (gemma-3-27b-it T3K functional)
- .github/workflows/vllm-nightly-tests-impl.yaml

### Checklist
APC: https://github.com/tenstorrent/tt-metal/actions/runs/17329991938 ✅ 

Single card frequent model: https://github.com/tenstorrent/tt-metal/actions/runs/17330003760 ❌  SDXL failure (unrelated) ✅ 
Single card demo: https://github.com/tenstorrent/tt-metal/actions/runs/17330029693 ❌ 1 failure (seen on main) ✅ 

T3K unit: https://github.com/tenstorrent/tt-metal/actions/runs/17330093562 ❌  1 failure - fixed on main ✅ 
T3K demo: https://github.com/tenstorrent/tt-metal/actions/runs/17330108740 ❌ 2 failures - both seen on main ✅
T3K perplexity: https://github.com/tenstorrent/tt-metal/actions/runs/17347821109

vLLM Nightly: https://github.com/tenstorrent/tt-metal/actions/runs/17330222243 ✅

handrewsTT committed:
squash flex ai commits

handrewsTT committed:
rebase integration changes

handrewsTT committed:
update submodules to main

handrewsTT committed:
undo rmsnorm changes

handrewsTT committed:
update gemma vision ccls

handrewsTT committed:
model runs

handrewsTT committed:
tidy up changes

handrewsTT committed:
gemma config for 4b

handrewsTT committed:
fix sliding attention

rdraskicTT committed:
Add gemma support

handrewsTT committed:
running model

handrewsTT committed:
simple_text_demo updates

handrewsTT committed:
parameterise enable_trace in simple_text_demo.py

handrewsTT committed:
split point change

handrewsTT committed:
test changes

handrewsTT committed:
move gemma into models/demos

handrewsTT committed:
add text demo

handrewsTT committed:
remove tests from TTT

handrewsTT committed:
tidy up

handrewsTT committed:
more tidy up

handrewsTT committed:
precommit and tidy up

handrewsTT committed:
fix license header

mtairum commented on models/tt_transformers/PERF.md:
```diff
@@ -45,6 +45,9 @@ This configuration uses bfp4 MLP and bfp8 attention weights for all models excep
 | Mistral-7B        | N150        | 95        | 99        | 29.75         | 100.24    |
 | Mistral-7B        | N300        | 95        | 99        | 47.01         | 65.95     |
 | Mistral-7B        | T3K         | 95        | 99        | 67.82         | 53.93     |
+| gemma-3-27b-it    | T3K         | 92        | 99        | 16.91         | 466.7     |
```
Since we're providing gemma3 own folder, you can put it there in a readme file.

Please make sure  in that readme that it mentions that this is an initial functional version of Gemma3, not focused on performance.

handrewsTT committed:
add gemma test to CI

handrewsTT committed:
add t3k demo tests

handrewsTT committed:
fix tests

handrewsTT committed:
single card func

handrewsTT committed:
add conftest

handrewsTT committed:
comment other tests and make mounts rw

handrewsTT committed:
rw nightly and t3k unit test

handrewsTT committed:
remove n300 flag for t3k unit test

handrewsTT committed:
N150 gemma3 fun

handrewsTT committed:
skip vision perf verification

handrewsTT committed:
base name for vision tests

handrewsTT committed:
gemma-3-4b ref output

handrewsTT committed:
gemma-3 perf file

handrewsTT committed:
update perfs

handrewsTT committed:
add acc check and uncomment other tests

handrewsTT committed:
add gemma accuracy tests

handrewsTT committed:
rebase submodules

handrewsTT committed:
tidy up workflows

handrewsTT committed:
demo and generator handle None image inputs

handrewsTT committed:
increase trace region size for vision demo

rdraskicTT committed:
Add vllm support for gemma3 (#27671)

### Ticket
https://github.com/tenstorrent/tt-metal/issues/22814

### Problem description
Provide context for the problem.

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [x] [vllm nightly
gemma](https://github.com/tenstorrent/tt-metal/actions/runs/17324656133)
- [ ] [vllm nightly all
models](https://github.com/tenstorrent/tt-metal/actions/runs/17327948991)

handrewsTT committed:
fix llama decoder

handrewsTT committed:
add qwen to vllm nightly workflow

handrewsTT committed:
read only vllm mount

mtairum commented on models/demos/gemma3/tests/test_ci_dispatch.py:
```diff
@@ -30,6 +28,24 @@ def test_ci_dispatch(model_weights):
+            "models/tt_transformers/tests/test_embedding.py",
+            "models/tt_transformers/tests/test_rms_norm.py",
+            "models/tt_transformers/tests/test_mlp.py",
+            # "models/tt_transformers/tests/test_attention.py",
+            # "models/tt_transformers/tests/test_attention_prefill.py",
```
Any reason why these are commented out?

mtairum commented on models/demos/gemma3/tests/test_mmp.py:
```diff
@@ -0,0 +1,102 @@
+)
+@pytest.mark.parametrize("device_params", [{"l1_small_size": 24576}], indirect=True)
+def test_multi_modal_inference(seq_len, batch_size, reset_seeds, device):
+    print("device:", device)
```
Remove print

mtairum commented on models/demos/gemma3/tests/test_mmp.py:
```diff
@@ -0,0 +1,102 @@
+    tt_output = tt_model(tt_input)
+
+    tt_output_torch = ttnn.to_torch(tt_output).squeeze(0)
+    print("reference_output ", reference_output.shape)
```
Remove prints

mtairum commented on models/demos/gemma3/tests/test_vision_attention.py:
```diff
@@ -0,0 +1,100 @@
+
+    tt_out = tt_model(attention_input)
+
+    print("TT output :", tt_out)
```
remove print

mtairum commented on models/demos/gemma3/tests/test_vision_cross_attention_transformer.py:
```diff
@@ -0,0 +1,101 @@
+    logger.info("Checking outputs")
+    out = ttnn.from_device(test_output)
+
+    print("out shape: ", out)
```
Remove prints

mtairum commented on models/demos/gemma3/tests/test_vision_layernorm.py:
```diff
@@ -0,0 +1,103 @@
+
+    logger.info("Compilation pass for LayerNorm")
+    tt_output = tt_model(tt_input)
+    print("tt_outputs ", tt_output)
```
Remove print

mtairum commented on models/demos/gemma3/tests/test_vision_mlp.py:
```diff
@@ -0,0 +1,88 @@
+
+    tt_output = tt_model(tt_input)
+
+    print("TT output shape:", tt_output)
```
Remove print

mtairum commented on models/demos/gemma3/tests/test_vision_pipeline.py:
```diff
@@ -0,0 +1,192 @@
+    logger.info("Checking outputs")
+    out = ttnn.from_device(test_output)
+    tt_output_torch = ttnn.to_torch(out, mesh_composer=ttnn.ConcatMeshToTensor(mesh_device, dim=0))[0, :, :, :]
+    print("reference_output ", reference_output.shape)
```
Remove print

mtairum commented on models/demos/gemma3/tt/gemma_vision_rmsnorm.py:
```diff
@@ -0,0 +1,172 @@
+We have handled the unit offset addition in the RMSNorm implementation directly into the TTNN Weights
+"""
+
+# SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
```
Copyright should be at top.

mtairum commented on models/tt_transformers/tt/common.py:
```diff
@@ -5,9 +5,11 @@
 from typing import Optional
 
 import torch
+from llama_models.llama3.api.datatypes import ImageMedia
```
For the future, does it make sense to rely on this llama3 API, since we're going for generic model support?

FIY @gwangTT

gwangTT replied:
Maybe something more generic like PILImage would be better?

mtairum commented on models/tt_transformers/tt/generator.py:
```diff
@@ -588,7 +652,7 @@ def prefill_forward(
 
     # Note: This function is called by vLLM
-    def decode_forward(
+    def decode_forward_llama_vision(
```
Should we start using a more generic name instead of llama vision? This is the path for gemma3 as well right?

Ok to leave for a followup PR.

handrewsTT committed:
add model README

mtairum commented on models/tt_transformers/tt/generator_vllm.py:
```diff
@@ -13,7 +13,7 @@
 
 import ttnn
-from models.tt_transformers.demo.simple_vision_demo import create_multimodal_model
+from models.demos.gemma3.demo.vision_demo import create_multimodal_model
```
Is this correct?

We shouldn't be referencing a specific external model from within TTT.
FIY @gwangTT

gwangTT replied:
Agreed. Maybe a good idea to make a separate generator_vllm.py file for Gemma?

mtairum commented on tests/scripts/single_card/run_single_card_demo_tests.sh:
```diff
@@ -49,6 +49,18 @@ run_qwen25_vl_func() {
 }
 
+run_gemma3_func() {
+  HF_MODEL=/mnt/MLPerf/tt_dnn-models/google/gemma-3-4b-it pytest models/demos/gemma3/demo/text_demo.py -k "ci-token-matching"
```
we might probably not run this token-matching for now, as it's prone to be changing very very frequently.

handrewsTT replied:
so should we just change this to something like `-k "performance and batch1"?

mtairum commented on tests/scripts/t3000/run_t3000_perplexity_tests.sh:
```diff
@@ -208,6 +208,22 @@ run_t3000_qwen3_perplexity_tests() {
+  echo "LOG_METAL: Running run_t3000_gemma3_accuracy_tests"
+  gemma3_27b=/mnt/MLPerf/tt_dnn-models/google/gemma-3-27b-it
+
+  HF_MODEL=$gemma3_27b pytest models/demos/gemma/demo/text_demo.py -k "ci-token-matching"
```
is this the config that runs the top1/top5 config on the text demo?

mtairum commented on models/demos/gemma3/README.md:
```diff
@@ -0,0 +1,55 @@
+
+## Performance
+
+Current model performance metrics can be found in [PERF.md](PERF.md)
```
wrong path. Should be `models/gemma3/PERF.md` or something like that.

handrewsTT committed:
add example to readme

handrewsTT committed:
remove prints and comments from tests

handrewsTT committed:
t3k perplexity path typo

gwangTT commented on .github/workflows/single-card-demo-tests-impl.yaml:
```diff
@@ -115,6 +115,10 @@ jobs:
           #          # Moved to t3k tests until OOM on single card runners resolved
                     # { name: "qwen7b", runner-label: "N300", performance: false, cmd: run_qwen7b_func, owner_id: U03PUAKE719}, # Mark O'Connor
                     { name: "qwen25_vl", runner-label: "N300", performance: true, cmd: run_qwen25_vl_func, owner_id: U07RY6B5FLJ},  #Gongyu Wang
+                    { name: "gemma3", runner-label: "N150", performance: true, cmd: run_gemma3_perf, owner_id: U08TJ70UFRT},  # Harry Andrews
+                    { name: "gemma3", runner-label: "N300", performance: true, cmd: run_gemma3_perf, owner_id: U08TJ70UFRT},  # Harry Andrews
+                    { name: "gemma3", runner-label: "N150", performance: false, cmd: run_gemma3_func, owner_id: U08TJ70UFRT},  # Harry Andrews
+                    { name: "gemma3", runner-label: "N300", performance: false, cmd: run_gemma3_func, owner_id: U08TJ70UFRT},  # Harry Andrews
```
I think for whole model tests, `performance` should be fine. If you are worried about whole model accuracy dropping, you could add a test of accuracy for the performance mode.

gwangTT replied:
If you have unit tests checking the PCCs of each module, there should be sufficient coverage on accuracy.

gwangTT commented on models/demos/qwen25_vl/tt/generator_vllm.py:
```diff
@@ -4,21 +4,21 @@
 from models.demos.qwen25_vl.tt.generator import Generator as QwenVLGenerator
 from models.demos.qwen25_vl.tt.model import DropInVisionTransformer, Transformer
 from models.demos.qwen25_vl.tt.model_config import VisionModelArgs
+from models.tt_transformers.tt.generator_vllm import input_processor_for_multimodal
```
Was this change intentional? tt_transformers' input_processor does not work for QwenVL

rdraskicTT committed:
Update README: vllm changes

rdraskicTT commented on .github/workflows/vllm-nightly-tests-impl.yaml:
```diff
@@ -67,19 +67,19 @@ jobs:
+            override-tt-config: "{\"data_parallel\": 8}",
             owner_id: U08GELBB1AP, # Ambrose Ling
           },
-          # { Removing this until #26734 is resolved
-          #   name: "[RB] Llama-3.1-8B-Instruct",
-          #   model: "meta-llama/Llama-3.1-8B-Instruct",
-          #   server-timeout: 35,
-          #   benchmark-timeout: 10,
-          #   runner-label: bh-rackbox,
-          #   mesh-device: P150x8,
-          #   tt-llama-text-ver: tt_transformers,
-          #   override-tt-config: "{\"data_parallel\": 8}",
-          #   owner_id: U08GELBB1AP, # Ambrose Ling
-          # },
```
Please put this under the comment again before merging

handrewsTT committed:
recomment vllm test and lower t3k acc

handrewsTT committed:
remove extra single card demo tests

handrewsTT committed:
update readme
